### ISSUE TITLE
Grid Using Sprites v2 example: margin=0 works differently in arcade==2.5.7 vs arcade==2.6.*

### ISSUE THREAD

**Issue by LJArendse** [pythonarcade/arcade#959]
## Documentation/Bug request:
I recently updated to the latest release version of arcade (arcade==2.6.2) from arcade==2.5.7. I am busy building a grid based game and noticed a change in the way the grid is '_rendering_' from version arcade==2.5.7. In short, in the _Grid Using Sprites v2 example_ setting `margin=0` works differently in `arcade==2.5.7` vs `arcade==2.6.*` (see https://api.arcade.academy/en/latest/examples/array_backed_grid_sprites_2.html#array-backed-grid-sprites-2). 

When setting `margin=0` in the _Grid Using Sprites v2 example_, in `arcade==2.5.7`, the grid is rendered smoothly as expected (i.e. there is no margin between the grid blocks/cells). However, when setting `margin=0` in the _Grid Using Sprites v2 example_, in `arcade==2.6.2`, the grid is rendered with a margin (i.e. there is a margin between the grid blocks/cells). 

The question I have is: How can I modify the code in _Grid Using Sprites v2 example_ such that I can render a smooth grid with no margin, in  arcade==2.6.2? 

### Actual behavior in arcade==2.6.2:
Setting the MARGIN=0 in _Grid Using Sprites v2 example_  results in:

<img width="449" alt="Screenshot 2021-09-19 at 11 00 15" src="https://user-images.githubusercontent.com/13453134/133922225-11adf848-23e1-4015-a1bf-bb574d69d4dd.png">

### Expected behavior in arcade==2.6.2:
Setting the MARGIN=0 in _Grid Using Sprites v2 example_  should, ideally for me, result in:

<img width="449" alt="Screenshot 2021-09-19 at 10 58 13" src="https://user-images.githubusercontent.com/13453134/133922235-452659df-de30-4f58-b292-7a8b9d4b0204.png">

### Steps to reproduce/example code:
1. Install `arcade==2.5.7`
2. Download the code from _Grid Using Sprites v2 example_ (https://api.arcade.academy/en/latest/examples/array_backed_grid_sprites_2.html#array-backed-grid-sprites-2)
3. Set the MARGIN=0, in line 24
```
MARGIN = 0
```
4. Run the _Grid Using Sprites v2 example_ code with the margin set to 0, using `arcade==2.5.7`. You should notice that the margin is not being rendered (we get a _smooth_ grid world).
5. Install `arcade==2.6.2`
6. Set the MARGIN=0, in line 24 (i.e. ensure you have done step 3)
```
MARGIN = 0
```
7. Run the _Grid Using Sprites v2 example_ code with the margin set to 0, using `arcade==2.6.2`. You should notice that the margin is now being rendered regardless of margin value.

or you can run the below code respectively in `arcade==2.5.7` and `arcade==2.6.2`, which already has the margin=0:
```python
"""
Array Backed Grid Shown By Sprites

Show how to use a two-dimensional list/array to back the display of a
grid on-screen.

This version makes a grid of sprites instead of numbers

If Python and Arcade are installed, this example can be run from the command line with:
python -m arcade.examples.array_backed_grid_sprites_2
"""
import arcade

# Set how many rows and columns we will have
ROW_COUNT = 15
COLUMN_COUNT = 15

# This sets the WIDTH and HEIGHT of each grid location
WIDTH = 30
HEIGHT = 30

# This sets the margin between each cell
# and on the edges of the screen.
MARGIN = 0

# Do the math to figure out our screen dimensions
SCREEN_WIDTH = (WIDTH + MARGIN) * COLUMN_COUNT + MARGIN
SCREEN_HEIGHT = (HEIGHT + MARGIN) * ROW_COUNT + MARGIN
SCREEN_TITLE = "Array Backed Grid Buffered Example"


class MyGame(arcade.Window):
    """
    Main application class.
    """

    def __init__(self, width, height, title):
        """
        Set up the application.
        """
        super().__init__(width, height, title)

        arcade.set_background_color(arcade.color.BLACK)

        # One dimensional list of all sprites in the two-dimensional sprite list
        self.grid_sprite_list = arcade.SpriteList()

        # This will be a two-dimensional grid of sprites to mirror the two
        # dimensional grid of numbers. This points to the SAME sprites that are
        # in grid_sprite_list, just in a 2d manner.
        self.grid_sprites = []

        # Create a list of solid-color sprites to represent each grid location
        for row in range(ROW_COUNT):
            self.grid_sprites.append([])
            for column in range(COLUMN_COUNT):
                x = column * (WIDTH + MARGIN) + (WIDTH / 2 + MARGIN)
                y = row * (HEIGHT + MARGIN) + (HEIGHT / 2 + MARGIN)
                sprite = arcade.SpriteSolidColor(WIDTH, HEIGHT, arcade.color.WHITE)
                sprite.center_x = x
                sprite.center_y = y
                self.grid_sprite_list.append(sprite)
                self.grid_sprites[row].append(sprite)

    def on_draw(self):
        """
        Render the screen.
        """

        # This command has to happen before we start drawing
        arcade.start_render()

        self.grid_sprite_list.draw()

    def on_mouse_press(self, x, y, button, modifiers):
        """
        Called when the user presses a mouse button.
        """

        # Change the x/y screen coordinates to grid coordinates
        column = int(x // (WIDTH + MARGIN))
        row = int(y // (HEIGHT + MARGIN))

        print(f"Click coordinates: ({x}, {y}). Grid coordinates: ({row}, {column})")

        # Make sure we are on-grid. It is possible to click in the upper right
        # corner in the margin and go to a grid location that doesn't exist
        if row < ROW_COUNT and column < COLUMN_COUNT:

            # Flip the location between 1 and 0.
            if self.grid_sprites[row][column].color == arcade.color.WHITE:
                self.grid_sprites[row][column].color = arcade.color.GREEN
            else:
                self.grid_sprites[row][column].color = arcade.color.WHITE


def main():
    MyGame(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)
    arcade.run()


if __name__ == "__main__":
    main()
```

**Comment by einarf** [comment]
Good catch. It seems there is a tiny difference in texture coordinates from last version. This only affects hidpi displays. This can be fixed by applying nearest interpolation when drawing your sprites

```py
        self.grid_sprite_list.draw(filter=self.ctx.NEAREST)
```

I'll try to dig a bit deeper because this is not optimal.

**Comment by einarf** [comment]
Figured it out. In 2.5 we did pad the borders of the atlas with a repeating pixel. Need to look into if this is still possible in an efficient way in the new atlas system.

**Comment by einarf** [comment]
Added repeating padding pixels around each texture in the atlas. This is what we did in 2.5.x to combat this exact issue.
![unknown](https://user-images.githubusercontent.com/5418180/134116944-18a8ee5b-0d01-49b0-950b-f134e3054dfd.png)

**Comment by einarf** [comment]
Working on retina so this should work for all hidps displays 👍 
<img width="562" alt="Screenshot 2021-09-21 at 08 06 22" src="https://user-images.githubusercontent.com/5418180/134120085-6ed02149-4c5f-479f-b4bf-9706cf9234c1.png">

**Comment by LJArendse** [comment]
Hi @einarf, thanks for all the help. This is awesome news, nicely done 💯  🥇 .
2ea45c2d9d397487ca695892322ac4095094cfc4 is a cool fix :sunglasses:

**Comment by einarf** [comment]
No problem! I need to optimize it later moving that part of the gpu, but the cost of this operation isn't too bad since we're resizing atlases on the gpu anyway. Textures are for the most part only added once.

**Comment by einarf** [comment]
Fixed in 2.6.3. Be free to reopen

**Comment by LJArendse** [comment]
Updated to version 2.6.3, everything is working perfectly now. Thanks for the help @einarf 😎 👍 .
